### PR TITLE
grad checkpointing

### PR DESCRIPTION
**Pull request by ethansmith2000**
# What does this PR do?

fixes to gradient checkpointing  based on this diagram
Untitled-19

I originally made these changes but then saw Transformer2DModel already had in the necessary changes in main! But the downblocks then need to have their checkpoint functions remove I believe which i have done in this commit

here are the differences in VRAM consumption on the LoRA training script you have provided for SDXL running at default parameters 
Before:
Screen Shot 2023-08-04 at 2 31 31 PM

After:

Screen Shot 2023-08-04 at 2 31 41 PM

Fixes # (issue)

improper gradient checkpointing that becomes problematic when Transformer2Dmodels consist of more than 1 BasicTransformerBlock

@patrickvonplaten @sayakpaul

### CODE REVIEW

**Comment by bghira** [comment]
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.125.06   Driver Version: 525.125.06   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA RTX A6000    On   | 00000000:56:00.0 Off |                  Off |
| 30%   49C    P2   112W / 300W |  43984MiB / 49140MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA RTX A6000    On   | 00000000:D6:00.0 Off |                  Off |
| 30%   49C    P2   116W / 300W |  43508MiB / 49140MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
+-----------------------------------------------------------------------------+
```

thanks. your fix, in combination with the earlier fix from today, makes it possible to also fine-tune the whole u-net on 48G of VRAM at 768x768.

**Comment by sayakpaul** [comment]
@ethansmith2000 thanks for your PR!

Could you please run `make fix-copies` from your `diffusers` fork and push the changes?

**Comment by patrickvonplaten** [comment]
@bghira I've just updated https://github.com/huggingface/diffusers/pull/4505 to "main" so that the initial gradient checkpointing fixes are included - could you double check if the PR with current main really leads to worse memory results?

**Comment by bghira** [comment]
oh, it absolutely does. it allowed me to actually rent 48G GPUs instead of 80G. i tested both, as it would have just been more convenient for me if it had worked with main.

haven't been using that trainer script yet, i have a local implementation i've been mashing together.

**Comment by patrickvonplaten** [comment]
Also asked on the PyTorch forum here: https://discuss.pytorch.org/t/gradient-checkpointing-inside-gradient-checkpointing/185861

**Comment by patrickvonplaten** [comment]
@bghira could you maybe post a reproducible of that form: https://github.com/huggingface/diffusers/pull/4474#issuecomment-1667486093 for your script - would love to reproduce the difference in memory here.

**Comment by bghira** [comment]
might take me time :) all my GPUs are training

**Comment by patrickvonplaten** [comment]
Ok after some internal discussion it seems like this "gradient checkpointing" inside "gradient checkpointing" works when using `use_reentrant=True`  is currently being worked on, but quite a novel feature (so probably doesn't work with PyTorch < 2). So it seems like it's not very robust right now. 

The advantage of allowing gradient checkpointing inside gradient checkpointing is that it allows us to build a more modular library where each component can have gradient checkpointing enabled. E.g. we ideally want both the `Transformer2DModel` and the unet blocks independently from each other be able to use gradient checkpointing. 

For now however the solution is this PR seems like the better, more robust way going forward - so let's go with it I'd say. I also don't see any real downsides to it :-)
